bndtools.core/src/org/bndtools/core/editors/quickfix/BuildpathQuickfixProcessor.java

```diff
@@ -0,0 +1,165 @@
+import org.eclipse.jdt.core.compiler.IProblem;
+import org.eclipse.jdt.core.dom.ASTNode;
+import org.eclipse.jdt.core.dom.ImportDeclaration;
+import org.eclipse.jdt.core.dom.Name;
+import org.eclipse.jdt.ui.text.java.IInvocationContext;
+import org.eclipse.jdt.ui.text.java.IJavaCompletionProposal;
+import org.eclipse.jdt.ui.text.java.IProblemLocation;
+import org.eclipse.jdt.ui.text.java.IQuickFixProcessor;
+
+import aQute.bnd.build.Container;
+import aQute.bnd.build.Project;
+import aQute.bnd.build.Workspace;
+import aQute.bnd.osgi.BundleId;
+import aQute.lib.exceptions.Exceptions;
+import bndtools.Plugin;
+import bndtools.central.Central;
+
+public class BuildpathQuickfixProcessor implements IQuickFixProcessor {
+
+	@Override
+	public boolean hasCorrections(ICompilationUnit unit, int problemId) {
+		switch (problemId) {
+			case IProblem.IsClassPathCorrect :
+				return true;
+			case IProblem.ImportNotFound :
+				return true;
+			case IProblem.UndefinedType :
+				return true;
+			default :
+				return false;
+		}
+	}
+
+	@Override
+	public IJavaCompletionProposal[] getCorrections(IInvocationContext context, IProblemLocation[] locations)
+		throws CoreException {
+		try {
+			List<IJavaCompletionProposal> proposals = new ArrayList<>();
+
+			ICompilationUnit compUnit = context.getCompilationUnit();
+			IJavaProject java = compUnit.getJavaProject();
+			if (java == null)
+				return null;
+
+			Project project = Central.getProject(java.getProject());
+			if (project == null)
+				return null;
+
+			boolean test = isInDir(project.getTestSrc(), compUnit.getResource());
+			Workspace workspace = project.getWorkspace();
+
+			for (IProblemLocation location : locations) {
+
+				if (!hasCorrections(context.getCompilationUnit(), location.getProblemId()))
+					return null;
+
+				String partialClassName = getPartialClassName(location.getCoveringNode(context.getASTRoot()));
+				if (partialClassName == null && location.getProblemArguments().length > 0)
+					partialClassName = location.getProblemArguments()[0];
+
+				if (partialClassName == null)
+					continue;
+
+				Map<String, List<BundleId>> result = workspace.search(partialClassName)
+					.orElseThrow(s -> new CoreException(new Status(IStatus.ERROR, Plugin.PLUGIN_ID, s)));
+
+				Set<BundleId> buildpath = getBundleIds(project.getBuildpath());
+				Set<BundleId> testpath = test ? getBundleIds(project.getTestpath()) : Collections.emptySet();
+
+				result.entrySet()
+					.forEach(e -> {
+						for (BundleId id : e.getValue()) {
+
+							if (test && !testpath.contains(id) && !buildpath.contains(id))
+								proposals.add(propose(e.getKey(), id, context, location, project, "-testpath"));
+
+							if (!buildpath.contains(id))
+								proposals.add(propose(e.getKey(), id, context, location, project, "-buildpath"));
+
+						}
+					});
+			}
+			return proposals.toArray(new IJavaCompletionProposal[0]);
+		} catch (Exception e) {
+			throw Exceptions.duck(e);
+		}
```

Comment:
Because this is the entry point from Eclipse into the quick fix processor, maybe it would be better to wrap this in a `CoreException` rather than allowing it to propagate up? I'm not 100% sure on Eclipse best practice here, but that seems to be what we've done elsewhere in Bndtools.

Reply:
If you look at the code you see that CoreException is thrown when we have something useful to tell, see line 85. All other exceptions are basically _out of contract_. No way to do anything useful there. If there are common exceptions that we want to handle we should change the API to use Result. 

In bnd, exceptions are treated as a boolean, they should just signal an out of contract failure. The new Result is really helpful here. Since not everybody follows that model, sometimes you need to catch an exception that contains information you cannot get in any other way. That is then necessary but we should try to treat exceptions as booleans, indicating you're in lala land, no way to known where you are anymore.

Reply:
Hmm, I'm not sure I follow here. If this was a Bnd internal function, changing the API to use Result would be a sensible option. But this is part of the Eclipse API - we don't have control over it, we have to try and comply with it.

The documentation for `IQuickFixProcessor.getCorrections()` states that we may throw `CoreException` if something goes wrong. It doesn't say that we're allowed to throw any other kind of exception, and if we do Eclipse's behaviour will be undefined (especially a checked exception masked as an unchecked exception). It might prevent other quick fix processors from running. Less likely, but more seriously it might crash our editor, or our entire Eclipse instance. Surely it is better to be a good citizen and do our best to make sure that our own internal errors (even unanticipated errors caused by bugs) don't do damage to any of our "neighbours"?

Reply:
The issue is that in 'normal' code the exception would always be a runtime exception. I decided from the beginning that an Checked Exceptions are evil and should be treated as runtime exceptions. I.e. if this exception is thrown, the code is wrong. The Exceptions class was added to treat the checked exceptions as RuntimeExceptions.

You're right about the consequences but an exception there means the code is fundamentally broken. I.e. like a NPE. It just should not happen and therefore should not have a decent  handling. It is Game Over.

Reply:
I get that an exception there means the code is fundamentally broken, that it should not happen. But if it *is* fundamentally broken/it does happen? What if we do cause an NPE, for example? Wouldn't it be better if we *didn't* bring down Eclipse or stop other quick fix processors from running?

If your answer to this is simply "no", I guess I will shrug my shoulders and note my disagreement for the record. I would sandbox it to prevent our contagion from spreading.

Reply:
But Eclipse is sandboxing it already ... Any runtime exception must be caught or Eclipse would not survive very long. It should also be logged in the error log, which Core Exceptions aren't I think. It happens to be visible here because I decided early on in bnd (1999 probably) that checked exceptions are evil and all methods are throwing Exception, basically treating all exceptions as runtime exceptions. The primary reason is to pass the original cause upwards instead of the long chains that too often miss the root cause. This strategy makes runtime exceptions visible in cases like this, cases that normal code can ignore. If this exception happens, it is like a core dump in C code or panic() in Rust. I want this to be really bad because it is a code error and must be fixed asap, not hidden in any way. The goal is to show the place where the thing failed and we should therefore not wrap it in any way. And I want it in the Eclipse log.

So far it worked fine for me. However, lambdas threw a bit of a wrench in this strategy and that makes us struggle a bit with it. I wish Java had an option to ignore checked exceptions, like every other language on earth. It is probably the worst idea since language design began. So due to lambdas, I'd probably not throw Exception on every method again. But that would increase the use of duck() because I refuse to play the game of checked exceptions unless forced by bad API design.

Reply:
If Eclipse is properly sandboxing it already, then scratch what I said - I agree that yours is the best implementation in that circumstance. Unnecessary wrapping of the exception makes it harder to find the root cause in the logs.